nixos/modules/programs/xonsh.nix

```diff
@@ -45,7 +45,23 @@ in
 
   config = mkIf cfg.enable {
 
-    environment.etc.xonshrc.text = cfg.config;
+    environment.etc.xonshrc.text = ''
+      # /etc/xonshrc: DO NOT EDIT -- this file has been generated automatically.
+
+
+      if not ''${...}.get('__NIXOS_SET_ENVIRONMENT_DONE'):
+          # The NixOS environment and thereby also $PATH
+          # haven't been fully set up at this point. But
+          # `source-bash` below requires `bash` to be on $PATH,
+          # so add one entry where it can be found for sure:
+          $PATH.add('/run/current-system/sw/bin')
```

Comment:
Is it OK to rely on `/run/current-system/sw/bin` containing some symlink to `bash`? Or should I use a path relative to a bash package, to force that to be pulled in as a dependency?

Reply:
I think this is sloppy: yes, bash is almost definitely going to be there but it's better to explicitely reference the bash store path if this module depends on bash (or a POSIX shell, if possible) to work.

Reply:
Thanks for the feedback!

> yes, bash is almost definitely going to be there but it's better to explicitely reference the bash store path if this module depends on bash [...] to work.

I've switched to `bin/` provided `pkgs.bash` in [`57a56afd`](https://github.com/NixOS/nixpkgs/compare/f581c64ffcd9226594e9a5975ff7ef2aae73969c..57a56afdd5ea589ffbd5fdb0eeef9638dc9e9601). As `source-bash` spawns a bash to source the file, then examines that bash session to adopt the environment (i.e. environment variables and their values) and aliases, I think `pkgs.bash` suffices and it doesn't have to be `pkgs.bashInteractive`, but I'm not quite sure I know all the differences between those attributes.

I've tested it in a VM and it seems to work fine.

> (or a POSIX shell, if possible)

I think that would be possible with [xonsh's `source-foreign`](https://xon.sh/aliases.html?highlight=source%20bash#source-foreign). Actually [`source-bash`](https://xon.sh/aliases.html?highlight=source%20bash#source-bash) is [just an alias for `source-foreign bash --sourcer=source`](https://github.com/xonsh/xonsh/blob/0.9.17/xonsh/aliases.py#L776). Though I'd have to investigate what options would have to be passed to `source-foreign` for a POSIX shell.

Is that something I should pursue?

Reply:
If bash is only needed to source the environment, `pkgs.bash` is the right one. `bashInteractive` is bash with documentation and readline support.

Reply:
Given it's called `source-bash` it would probably be wise so give it bash. I you want to try a strictly POSIX-compliant shell there is `pkgs.dash`.

Reply:
> Given it's called `source-bash` it would probably be wise so give it bash.

Well, I'd have used `source-foreign` directly for non-bash shell, not `source-bash`.

> I you want to try a strictly POSIX-compliant shell there is `pkgs.dash`.

> I think that would be possible with [xonsh's `source-foreign`](https://xon.sh/aliases.html?highlight=source%20bash#source-foreign).

It seems that only works for shell commands [known to xonsh](https://github.com/xonsh/xonsh/blob/da7bd1105b301c279a9798a976d083eb7c41a2e9/xonsh/foreign_shells.py#L92-L103), which currently are only
 - `bash`
 - `/bin/bash`
 - `zsh`
 - `/bin/zsh`
 - `/usr/bin/zsh`
 - `cmd`
 - `cmd.exe`

Everything else leads to a `KeyError` at [`xonsh/foreign_shells.py`, line 244](https://github.com/xonsh/xonsh/blob/0.9.17/xonsh/foreign_shells.py#L244).

Reply:
Thus I think we have to stick to Bash here for now.